File: skript-parser-java-impl/src/main/java/io/github/syst3ms/skriptparser/registration/TypeManager.java
package io.github.syst3ms.skriptparser.registration;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;
import java.util.regex.Matcher;

/**
 * Manages all different types.
 */
public class TypeManager {
    public static final String NULL_REPRESENTATION = "<none>";
    public static final String EMPTY_REPRESENTATION = "<empty>";
    private static final TypeManager instance = new TypeManager();
    private Map<String, Type<?>> nameToType = new HashMap<>();
    private Map<Class<?>, Type<?>> classToType = new HashMap<>();

    private TypeManager(){}

    public static TypeManager getInstance() {
        return instance;
    }

    public Map<Class<?>, Type<?>> getClassToTypeMap() {
        return classToType;
    }

    /**
     * Gets a {@link Type} by its exact name (the baseName parameter used in {@link Type#Type(Class, String, String)})
     * @param name the name to get the Type from
     * @return the corresponding Type, or {@literal null} if nothing matched
     */
    public Type<?> getByExactName(String name) {
        return nameToType.get(name);
    }

    /**
     * Gets a {@link Type} using {@link Type#syntaxPattern}, which means this matches any alternate and/or plural form.
     * @param name the name to get a Type from
     * @return the matching Type, or {@literal null} if nothing matched
     */
    public Type<?> getByName(String name) {
        for (Type<?> t : nameToType.values()) {
            Matcher m = t.getSyntaxPattern().matcher(name);
            if (m.matches()) {
                return t;
            }
        }
        return null;
    }

    /**
     * Gets a {@link Type} from its associated {@link Class}.
     * @param c the Class to get the Type from
     * @param <T> the underlying type of the Class and the returned Type
     * @return the associated Type, or {@literal null}
     */
    @SuppressWarnings("unchecked")
    public <T> Type<T> getByClassExact(Class<T> c) {
        return (Type<T>) classToType.get(c);
    }

    public <T> Type<? super T> getByClass(Class<T> c) {
        Type<? super T> type;
        do {
            Class<? super T> superclass = c.getSuperclass();
            type = getByClassExact(superclass);
        } while (type == null);
        return type;
    }

    public String toString(Object... objects) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < objects.length; i++) {
            if (i > 0) {
                sb.append(i == objects.length - 1 ? " and " : ", ");
            }
            Object o = objects[i];
            if (o == null) {
                sb.append(NULL_REPRESENTATION);
                continue;
            }
            Type<?> type = getByClass(o.getClass());
            if (type == null) {
                sb.append(Objects.toString(o));
            } else {
                sb.append(type.getToStringFunction().apply(o));
            }
        }
        return sb.length() == 0 ? EMPTY_REPRESENTATION : sb.toString();
    }

    /**
     * Gets a {@link PatternType} from a name. This determines the number (single/plural) from the input.
     * If the input happens to be the base name of a type, then a single PatternType (as in "not plural") of the corresponding type is returned.
     * @param name the name input
     * @return a corresponding PatternType, or {@literal null} if nothing matched
     */
    public PatternType<?> getPatternType(String name) {
        if (nameToType.containsKey(name)) { // Might as well avoid the for loop in this case
            return new PatternType<>(nameToType.get(name), false);
        }
        for (Type<?> t : nameToType.values()) {
            Matcher m = t.getSyntaxPattern().matcher(name);
            if (m.matches()) {
                String pluralGroup = m.group("plural");
                return new PatternType<>(t, pluralGroup == null || pluralGroup.isEmpty());
            }
        }
        return null;
    }

    void register(SkriptRegistration reg) {
        for (Type<?> type : reg.getTypes()) {
            nameToType.put(type.getBaseName(), type);
            classToType.put(type.getTypeClass(), type);
        }
    }
}

File: skript-parser-kotlin/src/main/kotlin/io/github/syst3ms/skriptparser/registration/TypeManager.kt
package io.github.syst3ms.skriptparser.registration

import io.github.syst3ms.skriptparser.classes.PatternType

@Suppress("UNCHECKED_CAST")
/**
 * Manages all different types.
 */
object TypeManager {
    private val nameToType = mutableMapOf<String, Type<out Any>>()
    private val classToType: MutableMap<Class<in Any>, Type<out Any>> = mutableMapOf()

    /**
     * Gets a [Type] by its exact name (the baseName parameter used in [Type]'s constructor])
     * @param name the name to get the Type from
     * @return the corresponding Type, or null if nothing matched
     */
    fun getByExactName(name: String): Type<*>? {
        return nameToType[name]
    }

    /**
     * Gets a [Type] using [Type.syntaxPattern], which means this matches any alternate and/or plural form.
     * @param name the name to get a Type from
     * @return the matching Type, or null if nothing matched
     */
    fun getByName(name: String): Type<out Any>? {
        return nameToType.values.firstOrNull { it.syntaxPattern matches name }
    }

    /**
     * Gets a [Type] from its associated [Class].
     * @param c the Class to get the Type from
     * @return the associated Type, or null
     */
    fun getByClass(c: Class<in Any>): Type<out Any>? {
        return classToType[c]
    }

    /**
     * Gets a [PatternType] from a name. This determines the number (single/plural) from the input.
     * If the input happens to be the base name of a type, then a single PatternType (as in "not plural") of the corresponding type is returned.
     * @param name the name input
     * @return a corresponding PatternType, or null if nothing matched
     */
    fun getPatternType(name: String): PatternType<*>? {
        if (nameToType.containsKey(name)) { // Might as well avoid the for loop in this case
            return PatternType((nameToType[name] as Type<Any>?)!!, false)
        }
        for (t in nameToType.values) {
            val m = t.syntaxPattern.matchEntire(name)
            if (m != null) {
                return PatternType(t, m.groups["plural"] == null)
            }
        }
        return null
    }

    internal fun register(reg: SkriptRegistration) {
        for (type in reg.types) {
            nameToType[type.baseName] = type
            classToType[type.c as Class<in Any>] = type
        }
    }
}